src/Warehouse.java

```diff
@@ -60,12 +64,33 @@ public List<Product> getProductsInStock() {
     public boolean isEmpty() {
         return devices.isEmpty();
     }
+    public double calculateAveragePurchasePriceByDeviceType(Devices deviceType) throws InvalidSearchFilterException {
+        double totalPurchasePrice = 0.0;
+        int numOfDevices = 0;
+
+        FilterByDevice deviceFilter = new FilterByDevice(deviceType);
+
+        for (Product product : devices) {
+                if (deviceFilter.isInFilter(product)) {
+                    totalPurchasePrice += product.getPurchasePrice();
+                    numOfDevices++;
+                } else {
+                    throw new InvalidSearchFilterException ("Invalid filter");
```

Comment:
isInFilter lancia già la stessa exeption di base, credo sia ripetitivo  rimetterlo perchè fa parte già dei controlli del filtro, toglierei quindi riga 77 e 78. poi si parte col secondo if che contiene la logica del metodo e li giustamente controlliamo prima se ci sono prodotti su cui applicare questa logica.Lancerei un'exeption a linea 82(forse anche un'ExecutionExeption dando come messaggio che non ci sono prodotti compatibili ) per poi chiudere il tutto come hai fatto .

Reply:
Come dice Silvio non ha scopo questa eccezione scritta poichè questo controllo già è fatto nel filtro. Quindi si linea 78 va eliminata. Il controllo che devi fare innanzitutto prima di ciclare è che il warehouse non sia vuoto.  (Andrebbe creata un eccezione al riguardo, questo controllo viene richiamato da diversi metodi. Lo aggiungiamo ai task poi) Una volta controllato che non sia vuoto allora puoi ciclare. Mentre cicli controlli se ci sono prodotti che rispettano il filtro e di conseguenza fai ciò che devi se ci sono prodotti, altrimenti se non ci sono prodotti ritorni nell'else il messaggio che hai scritto nel secondo if e ritorni anche la media che quindi è 0.0.  Sostanzialmente quel secondo if a riga 83 diventa il tuo else a riga 78.

Reply:
Allora, ho cercato di seguire alla lettera le vostre istruzioni. Gabriel mi diceva: “ Mentre cicli controlli se ci sono prodotti che rispettano il filtro e di conseguenza fai ciò che devi se ci sono prodotti, altrimenti se non ci sono prodotti ritorni nell'else il messaggio che hai scritto nel secondo if e ritorni anche la media che quindi è 0.0.  Sostanzialmente quel secondo if a riga 83 diventa il tuo else a riga 78.”. il problema è che quell’else è all’interno del ciclo che analizza tutti i device nella lista quindi qualora il primo elemento della lista non fosse un device del tipo cercato il codice ritornerebbe il messaggio che non ci sono device di quel tipo e sarebbe un errore, questo messaggio va lanciato dopo che il ciclo finisce, quindi ho mantenuto la versione di Silvio. Per quanto riguarda la Execution Exception che diceva Silvio, se la lancio il programma si arresta e non mi permette di ritornare la media di 0.0 qualora non ci fossero dispositivi del tipo cercato, quindi ho lasciato semplicemente il messaggio stampato come avevo già fatto. Per il resto ho modificato il codice come specificato in review.